internal/contentmanager/source/dynamicsource.go

```diff
@@ -0,0 +1,223 @@
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/informers"
+	cgocache "k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	source "github.com/operator-framework/operator-controller/internal/contentmanager/source/internal"
+)
+
+type DynamicSourceConfig struct {
+	// DynamicInformerFactory is the dynamicinformer.DynamicSharedInformerFactory
+	// that is used to generate the informers configured on this sources startup.
+	// If you use a dynamicinformer.DynamicSharedInformerFactory that you've
+	// used previously, it must not have been used to start a new informer for
+	// the same GVR. You can not start or configure an informer that has already
+	// been started, even after it has been stopped. Reusing an informer factory
+	// may result in attempting to configure and start an informer that has
+	// already been started.
+	DynamicInformerFactory dynamicinformer.DynamicSharedInformerFactory
+	// GVR is the GroupVersionResource that this source is responsible
+	// for creating and configuring an informer for
+	GVR schema.GroupVersionResource
+	// Owner is the client.Object that owns the managed content that this
+	// source will be creating an informer to react to events for. This
+	// field is used to attempt to requeue the owning client.Object for
+	// reconciliation on a watch error after a previously successful sync
+	Owner client.Object
+	// Handler is the handler.EventHandler that is used to react to events
+	// received by the configured source
+	Handler handler.EventHandler
+	// Predicates are the predicate.Predicate functions used to determine
+	// if a triggered event should be reacted to
+	Predicates []predicate.Predicate
+	// OnPostSyncError is the callback function that is used when the source
+	// initially synced successfully and later encountered an error
+	OnPostSyncError func(context.Context)
+}
+
+func NewDynamicSource(cfg DynamicSourceConfig) *dynamicInformerSource {
+	return &dynamicInformerSource{
+		cfg:        cfg,
+		errChan:    make(chan error, 1),
+		syncedChan: make(chan struct{}, 1),
+	}
+}
+
+// dynamicInformerSource is an implementation of the
+// ReaderCloserSyncingSource interface. It is used
+// to create an informer, using the provided dynamic informer
+// factory, for the configured GroupVersionResource.
+//
+// The informer is configured with a WatchEventErrorHandler that
+// stops the informer, and if it had previously synced successfully
+// it attempts to requeue the provided Owner for reconciliation and
+// calls the provided OnWatchError function.
+type dynamicInformerSource struct {
+	cfg        DynamicSourceConfig
+	errOnce    sync.Once
+	errChan    chan error
```

Comment:
```suggestion
erroredChan chan struct{}
```

And the errOnce could be:
```go
			dis.err = err
			close(dis.errChan)
```

And _then_ in WaitForSync:
```go
	case <-dis.erroredChan:
		return dis.err
```

Reply:
addressed in https://github.com/operator-framework/operator-controller/pull/1119/commits/af075c5137065a5e0b2e0cfc1b4debbb367f792f